### PR TITLE
Core/Spell: Fixed serverside creature orientation while casting

### PR DESCRIPTION
**Pull request by Golrag**
Closes https://github.com/TrinityCore/TrinityCore/issues/2947
Any feedback is welcome ! :)

### CODE REVIEW

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -12764,10 +12764,7 @@ Unit* Creature::SelectVictim()
         return NULL;
 
     if (target && _IsTargetAcceptable(target) && CanCreatureAttack(target))
-    {
-        SetInFront(target);
         return target;
```
Comment (joschiwald) [inline src/server/game/Entities/Unit/Unit.cpp]: does it work without trouble? is serverside orientation still updated correctly?

Comment (Golrag) [inline src/server/game/Entities/Unit/Unit.cpp]: Yes, it is handled here: https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp#L180
note: some spells should allow NPC's to walk/attack while channeling (like bone flurry) but this doesn't work (yet)

Comment (joschiwald) [inline src/server/game/Entities/Unit/Unit.cpp]: movementgenerator update is blocked while casting (and channeling), is serverside orientation updated correctly while casting?

Comment (Golrag) [inline src/server/game/Entities/Unit/Unit.cpp]: Good question, will test this asap.. Shouldn't SPELL_ATTR0_CAST_TRACK_TARGET take care of this ?

Comment (Golrag) [inline src/server/game/Entities/Unit/Unit.cpp]: Just tested it with a simple .cast back 11366, and it is not (spell was cast but serverside the npc didn't change his orientation until the cast was done).
But, if a player is tanking an NPC and NPC cast spell on another player, without this change it would have had his serverside orientation looking at the tank.

Comment (joschiwald) [inline src/server/game/Entities/Unit/Unit.cpp]: > But, if a player is tanking an NPC and NPC cast spell on another player, without this change it would have had his serverside orientation looking at the tank.

and exactly this is the issue of #10803

Comment (Golrag) [inline src/server/game/Entities/Unit/Unit.cpp]: Maybe set caster in front of target in spell:update like I did in this PR but with other spells ? I just don't know which spells need to track their focustarget :/
Will take a look at this

File: src/server/game/Entities/Creature/Creature.h [hunk 2]
```diff
@@ -659,7 +659,7 @@ class Creature : public Unit, public GridObject<Creature>, public MapObject
 
         // Handling caster facing during spellcast
         void SetTarget(ObjectGuid guid) override;
-        void FocusTarget(Spell const* focusSpell, WorldObject const* target);
+        void FocusTarget(Spell const* focusSpell, ObjectGuid targetGuid);
```
Comment (joschiwald) [inline src/server/game/Entities/Creature/Creature.h]: pls revert this change, or explain why this is needed

Comment (Golrag) [inline src/server/game/Entities/Creature/Creature.h]: To allow focus on "nothing" with ObjectGuid::Empty
I can change it back and use ObjectGuid::Empty if target does not exist if you prefer that :)

Comment (Warpten) [inline src/server/game/Entities/Creature/Creature.h]:
```c++
To allow focus on "nothing" with ObjectGuid::Empty
```

This doesnt make sense

Comment (Golrag) [inline src/server/game/Entities/Creature/Creature.h]: Should I just remove the creatures target then ?

Comment (Golrag) [inline src/server/game/Entities/Creature/Creature.h]: Some creatures stop targetting while casting example: http://www.wowhead.com/npc=72762 when casting http://www.wowhead.com/spell=147310/gust-of-wind

Comment (joschiwald) [inline src/server/game/Entities/Creature/Creature.h]: the wotlk one is (should be) handled by SPELL_ATTR5_DONT_TURN_DURING_CAST

> Should I just remove the creatures target then ?

don't call FocusTarget if caster == target like it is used in other places

Comment (Golrag) [inline src/server/game/Entities/Creature/Creature.h]: If caster == target FocusTarget will be called but with no target now. So _focusSpell will have a value. Will fix that asap, still one question: Should I call FocusTarget when the spell does not have a target ? If I do, when the spell is finished ReleaseFocus will be called and creature will start attack the tank again, If I don't.. ehm he won't I guess, so I will need to look at that :p

Comment (joschiwald) [inline src/server/game/Entities/Creature/Creature.h]: > Should I call FocusTarget when the spell does not have a target ?

i don't see any need to focus nothing

Comment (Golrag) [inline src/server/game/Entities/Creature/Creature.h]: Updated, instead of calling focusTarget with no target, I removed the creatures target when it has no target or when target is self

**Comment by Golrag** [comment]
PR by @Treeston https://github.com/TrinityCore/TrinityCore/pull/15641
